subsys/shell/shell_mqtt.c

```diff
@@ -0,0 +1,659 @@
+
+	ctx->fds[0].events = ZSOCK_POLLIN;
+	ctx->nfds = 1;
+}
+
+static void clear_fds(struct shell_mqtt *ctx)
+{
+	ctx->nfds = 0;
+}
+
+static int wait(int timeout)
+{
+	int ret = 0;
+
+	if (sh_mqtt->nfds > 0) {
+		ret = zsock_poll(sh_mqtt->fds, sh_mqtt->nfds, timeout);
+		if (ret < 0) {
+			LOG_ERR("poll error: %d", errno);
+		}
+	}
+
+	return ret;
+}
+
+static void mqtt_connect_work(struct k_work *work);
+
+/*
+ * Work routine to process MQTT packet and keep alive MQTT connection
+ */
+static void mqtt_process_work(struct k_work *work)
+{
+	int64_t remaining = APP_SLEEP_MSECS;
+	int64_t start_time = k_uptime_get();
+	struct shell_mqtt *ctx = NULL;
+	int rc;
+
+	ctx = CONTAINER_OF(work, struct shell_mqtt, mqtt_work);
+
+	/* Listen to the port for 500 ms as defined by APP_SLEEP_MSECS */
+	while (remaining > 0) {
+		if (wait(remaining)) {
+			rc = mqtt_input(&ctx->client_ctx);
+			if (rc != 0) {
+				LOG_ERR("mqtt_input error: %d", rc);
+			}
+		}
+
+		rc = mqtt_live(&ctx->client_ctx);
+		if (rc != 0 && rc != -EAGAIN) {
+			LOG_ERR("mqtt_live error: %d", rc);
+			return;
+		} else if (rc == -ENOTCONN) {
+			(void)mqtt_disconnect(&ctx->client_ctx);
+			k_work_init_delayable(&ctx->mqtt_work, mqtt_connect_work);
+			(void)k_work_reschedule(&ctx->mqtt_work, K_SECONDS(MQTT_PROCESS_INTERVAL));
+			return;
+		} else if (rc == 0) {
+			rc = mqtt_input(&ctx->client_ctx);
+			if (rc != 0) {
+				LOG_ERR("mqtt_input error: %d", rc);
+			}
+		}
+
+		remaining = APP_SLEEP_MSECS + start_time - k_uptime_get();
+	}
+
+	/* Reschedule the work */
+	k_work_init_delayable(&ctx->mqtt_work, mqtt_process_work);
+	(void)k_work_reschedule(&ctx->mqtt_work, K_SECONDS(MQTT_PROCESS_INTERVAL));
+}
+
+static int mqtt_subscribe_config(struct mqtt_client *const client)
+{
+	/* subscribe to config information */
+	struct mqtt_topic subs_topic = {
+		.topic = {
+			.utf8 = sh_mqtt->sub_topic,
+			.size = strlen(sh_mqtt->sub_topic)
+		},
+		.qos = MQTT_QOS_1_AT_LEAST_ONCE
+	};
+	const struct mqtt_subscription_list subs_list = {
+		.list = &subs_topic,
+		.list_count = 1U,
+		.message_id = 1U
+	};
+	int ret;
+
+	ret = mqtt_subscribe(client, &subs_list);
+	if (ret) {
+		LOG_ERR("Failed to subscribe to %s item, error %d", subs_topic.topic.utf8, ret);
+	}
+
+	return ret;
+}
+
+static void mqtt_subscribe_work(struct k_work *work)
+{
+	struct shell_mqtt *ctx = NULL;
+	int ret;
+
+	ctx = CONTAINER_OF(work, struct shell_mqtt, mqtt_work);
+
+	ret = mqtt_subscribe_config(&ctx->client_ctx);
+	if (ret == 0) {
+		/* Change the work to process/keepalive */
+		LOG_DBG("Scheduling MQTT process & keepalive work");
+		k_work_init_delayable(&ctx->mqtt_work, mqtt_process_work);
+		(void)k_work_reschedule(&ctx->mqtt_work, K_SECONDS(MQTT_PROCESS_INTERVAL));
+	} else {
+		(void)mqtt_abort(&ctx->client_ctx);
+		k_work_init_delayable(&ctx->mqtt_work, mqtt_connect_work);
+		(void)k_work_reschedule(&ctx->mqtt_work, K_SECONDS(MQTT_PROCESS_INTERVAL));
+	}
+}
+
+static void mqtt_connect_work(struct k_work *work)
```

Comment:
@rlubos is it fine to do this in a system work queue?

Reply:
It's also a blocking call (it needs to wait for a response), these shall be avoided on a system workqueue in general (note that if the work that is currently executed blocks, the system has no means to reschedule to a next work, the entire workqueue is blocked then). 

If you don't want to change the design of this module, and keep thich work-based approach, I think it'd be easiest to simply create a separate workqueue for this library.

Reply:
How about the memory footprint, i.e. separate workq versus thread?

Reply:
I don't have the exact number, I think it should be pretty close, I'd expect sligtly more for a workqueue. It's just a thread with some extra logic on top (the code is already compiled in due to sys workq).

Reply:
Got it, thanks